### PR TITLE
User can see the start and end of the puzzle.

### PR DESCRIPTION
**Pull request by danielduffield**
Created a start and end point that randomly generate each time the game is started, with a distance no less than 4 between them.

### CODE REVIEW

**Comment by danielduffield** [comment]
screen shot 2017-06-14 at 3 26 11 pm

File: main.js [hunk 1]
```diff
@@ -17,6 +18,62 @@ function generateTiles() {
...
+      if (isHidden(board[i][j]) && !isHiddenCorner(board[i][j])) {
+        candidates.push(board[i][j])
+      }
+    }
+  }
+  return candidates
+}
+
+function shuffleArray(array) {
+  for (let i = array.length - 1; i > 0; i--) {
```
Comment (thebearingedge) [inline main.js]: Try not to mutate your inputs.

File: main.js [hunk 2]
```diff
@@ -48,6 +105,14 @@ function renderTile(tile, rowNum) {
   if (tile.isSelected === true) {
     $tile.classList.add('selected')
   }
+  if (tile.goal) {
```
Comment (thebearingedge) [inline main.js:109]: Try to simplify this goal/classlist thing.

File: main.js [hunk 3]
```diff
@@ -78,40 +141,47 @@ function swapTiles(coordinates) {
...
-      board = swapTiles(selectedTiles)
-      selectedTiles = []
-    }
-    let $board = document.getElementById('board-render')
-    $board.removeEventListener('click', selectTile)
-    document.getElementById('game-board').removeChild($board)
-    $board = renderBoard(board)
-    document.getElementById('game-board').appendChild($board)
-    $board.addEventListener('click', selectTile)
+  if ((event.target.classList[1] === 'hidden-tile') || !(event.target.classList[0] === 'board-tile')) {
```
Comment (thebearingedge) [inline main.js]: Can you name this expression?
```js
function isApplicable(target) {
  // check the target, return boolean
}

**Review by thebearingedge — CHANGES_REQUESTED** [review]

File: main.js [hunk 4]
```diff
@@ -39,11 +39,12 @@ function getGoalCandidates(board) {
 }
 
 function shuffleArray(array) {
-  for (let i = array.length - 1; i > 0; i--) {
+  let shuffled = array.splice()
```
Comment (thebearingedge) [inline main.js]: Double-check [`splice`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice?v=example) vs [`slice`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice?v=example). Make sure you understand how these unfortunately-named functions work 😄 
Then you are approved to merge.

**Review by thebearingedge — CHANGES_REQUESTED** [review]

File: main.js [hunk 5]
```diff
@@ -17,6 +18,63 @@ function generateTiles() {
...
+      if (isHidden(board[i][j]) && !isHiddenCorner(board[i][j])) {
+        candidates.push(board[i][j])
+      }
+    }
+  }
+  return candidates
+}
+
+function shuffleArray(array) {
+  let shuffled = array.splice()
```
Comment (thebearingedge) [inline main.js]: Double-check [`splice`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice?v=example) vs [`slice`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice?v=example). Make sure you understand how these unfortunately-named functions work 😄 
Then you are approved to merge.

**Review by thebearingedge — CHANGES_REQUESTED** [review]
Double-check [`splice`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice?v=example) vs [`slice`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice?v=example). Make sure you understand how these unfortunately-named functions work 😄 
Then you are approved to merge.

**Pull request merged**